user_tasks/models.py

```diff
@@ -0,0 +1,205 @@
+
+    def increment_completed_steps(self, steps=1):
+        """
+        Increase the value of :py:attr:`completed_steps` by the given number and save, then check for cancellation.
+
+        If cancellation of the task has been requested, a TaskCanceledException
+        will be raised to abort execution.  If any special cleanup is required,
+        this exception should be caught and handled appropriately.
+
+        This method should be called often enough to provide a useful
+        indication of progress, but not so often as to cause undue burden on
+        the database.
+        """
+        UserTaskStatus.objects.filter(pk=self.id).update(completed_steps=F('completed_steps') + steps,
+                                                         modified=now())
+        self.refresh_from_db(fields={'completed_steps', 'modified', 'state'})
+        if self.parent:
+            self.parent.increment_completed_steps(steps)  # pylint: disable=no-member
+        # Was a cancellation command recently sent?
+        if self.state == self.CANCELED and not self.is_container:
+            raise TaskCanceledException
+
+    def increment_total_steps(self, steps):
+        """Increase the value of :py:attr:`total_steps` by the given number and save."""
+        # Assume that other processes may be making concurrent changes
+        UserTaskStatus.objects.filter(pk=self.id).update(total_steps=F('total_steps') + steps, modified=now())
+        self.refresh_from_db(fields={'total_steps', 'modified'})
+        if self.parent:
+            self.parent.increment_total_steps(steps)  # pylint: disable=no-member
+
+    def cancel(self):
+        """Cancel the associated task if it hasn't already finished running."""
+        if self.is_container:
+            children = UserTaskStatus.objects.filter(parent=self)
+            for child in children:
+                child.cancel()
+        elif self.state in (UserTaskStatus.PENDING, UserTaskStatus.RETRYING):
+            current_app.control.revoke(self.task_id)
+        with transaction.atomic():
+            status = UserTaskStatus.objects.select_for_update().get(pk=self.id)
+            if status.state in (UserTaskStatus.CANCELED, UserTaskStatus.FAILED, UserTaskStatus.SUCCEEDED):
+                return
+            status.state = UserTaskStatus.CANCELED
+            status.save(update_fields={'state', 'modified'})
+            self.state = status.state
+            self.modified = status.modified
+
+    def fail(self, message):
+        """
+        Mark the task as having failed for the given reason, and save it.
+
+        The message will be available as the :py:attr:`UserTaskArtifact.text`
+        field of a UserTaskArtifact with name "Error".  There may be more than
+        one such artifact for a single UserTaskStatus, especially if it
+        represents a container for multiple parallel tasks.
+        """
+        UserTaskArtifact.objects.create(status=self, name='Error', text=message)
```

Comment:
Given that some operations (like this one) make bottom-up changes in the status record hierarchy and others (like canceling a task) go top-down, I'm a bit concerned about deadlocks where transactions span more than a single record in the chain.  Suppose I could put one around the artifact creation and state update, at least.